src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -143,23 +143,43 @@ public class NetworkPeerConnection : IDisposable
         /// <summary>Task responsible for reading incoming messages from the stream.</summary>
         private Task receiveMessageTask;
 
+        /// <summary>Queue of incoming messages distributed to message consumers.</summary>
+        public MessageProducer<IncomingMessage> MessageProducer { get; private set; }
+
+        /// <summary>Consumer of messages coming from connected clients.</summary>
+        private readonly EventLoopMessageListener<IncomingMessage> messageListener;
+
+        /// <summary>New network state to set to the peer when shutdown is initiated.</summary>
+        private NetworkPeerState setPeerStateOnShutdown;
+
         /// <summary>
         /// Initializes an instance of the object.
         /// </summary>
         /// <param name="peer">Network peer the node is connected to.</param>
         /// <param name="client">Connected network client to the peer.</param>
+        /// <param name="messageReceivedCallback">Callback to be called when a new message arrives from the peer.</param>
         /// <param name="dateTimeProvider">Provider of time functions.</param>
         /// <param name="loggerFactory">Factory for creating loggers.</param>
-        public NetworkPeerConnection(NetworkPeer peer, NetworkPeerClient client, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+        public NetworkPeerConnection(NetworkPeer peer, NetworkPeerClient client, Func<IncomingMessage, Task> messageReceivedCallback, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
         {
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{peer.PeerAddress.Endpoint}] ");
             this.dateTimeProvider = dateTimeProvider;
 
+            this.setPeerStateOnShutdown = NetworkPeerState.Offline;
             this.Peer = peer;
             this.Client = client;
             this.CancellationSource = new CancellationTokenSource();
-            this.cancelRegistration = this.CancellationSource.Token.Register(this.InitiateShutdown);
+
+            // When the cancellation source is cancelled, the registered callback is executed within 
+            // the context of the thread that invoked the cancellation. However, we need InitiateShutdown to be 
+            // called in separation of that to avoid deadlock.
+            this.cancelRegistration = this.CancellationSource.Token.Register(() => Task.Run(() => this.InitiateShutdown()));
```

Comment:
This means you lose the task of the shutdown processes.
If an exception happens in the `InitiateShutdown` where does it propagate?

Reply:
this is reworked in part 18 as i have learned a lot about cancellation tokens over the past weeks

yes, you would lose the task, but the dispose code is depending on it, so if it crashed, you'd be fked up (node would crash), but because of that dependency, it is not needed to hold a reference to that task. anyway, reworked in p18